### PR TITLE
Gh 218 controller and auth (merge in first)

### PR DESCRIPTION
**Pull request by rocky341**
Hi @t92549 
This is a bit of a big PR sorry :/ 

Includes integration of the controller directly into the rest client which removes the need for CRDs.

Also includes a new auth profile setup.

### CODE REVIEW

File: gaffer-as-a-service/gaas-rest/worker/README.md [hunk 1]
```diff
@@ -0,0 +1 @@
+The dockerfile in this folder is for the worker which installs Gaffer instances.
```
Comment (t92549) [inline gaffer-as-a-service/gaas-rest/worker/README.md:1]: A better README might be nice

File: gaffer-as-a-service/pom.xml [hunk 2]
```diff
@@ -30,15 +30,14 @@
...
-        <module>gaffer-controller</module>
         <module>gaas-rest</module>
     </modules>
 
     <properties>
         <gaffer.version>1.17.1</gaffer.version>
         <kubernetes.version>14.0.0</kubernetes.version>
         <springboot.version>2.5.8</springboot.version>
-        <fasterxml.jackson.version>2.11.2</fasterxml.jackson.version>
+        <fasterxml.jackson.version>2.9.8</fasterxml.jackson.version>
```
Comment (t92549) [inline gaffer-as-a-service/pom.xml:40]: Jackson downgrade?

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java [hunk 3]
```diff
@@ -25,23 +27,20 @@
...
-import static uk.gov.gchq.gaffer.common.util.Constants.VERSION;
-import static uk.gov.gchq.gaffer.gaas.factories.GaaSGraphsFactory.from;
+
 import static uk.gov.gchq.gaffer.gaas.factories.GaaSRestExceptionFactory.from;
-import static uk.gov.gchq.gaffer.gaas.util.Properties.NAMESPACE;
 
 @Repository
-public class CRDClient {
-    private static final Logger LOGGER = LoggerFactory.getLogger(CRDClient.class);
+public class GafferClient {
```
Comment (t92549) [inline gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java:41]: Just so I am getting my head round this, the way it used to work was that the rest api would make a custom resource definition (crd) request which basically says make me this custom Gaffer. Then the CRDClient would be listening for that request and actually go about using kubernetes to create it? But now, the CRDClient is all binned and the rest api itself calls kubernetes to create the Gaffers.

I know it's a little late to be asking questions like this but just wondering why this move was decided? Is it as secure? And if it is the best way to do it, why was it done with a CRD Client in the first place? If I remember the initial Kai did something more akin to this but then it was reworked to use CRD Client and now it seems to be going back again to a slightly simpler form (simpler is probably better!).

Sorry for such a big question!

Comment (t92549) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java:41]: And to do this it is using `io.fabric8.kubernetes.client.KubernetesClient`? Is it official? How widely used / trusted is this? And if it is not official, can the official api not be used to do this?

**Review by t92549 — CHANGES_REQUESTED** [review]
Thanks for the PR! It is big but also hard to create such a huge change without a huge PR!
I have mostly asked questions but tagged as changes requested for the small things like copyright dates, sorry to be picky!

File: gaas-ui/README.md [hunk 4]
```diff
@@ -87,6 +87,14 @@ Make sure you are in the UI directory when running the commands.
 * `docker logs -f [container ID]` shows a log of the actions happening to the UI container.
 
 Once the process has finished, visit (http://localhost:8080), where you will be able to see the UI.
+## Podman and Docker Hub
+* `podman login docker.io` Logs you in to docker hub using Podman, enter your docker hub username and password.
+* `podman machine init` Initialises the Podman VM.
+* `podman machine start` Starts the Podman VM.
+* `podman build -t docker.io/YOUR-DOCKER-HUB-USERNAME/ui:latest .` builds a production version of Kai as a container.
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaas-ui/README.md:94]: Kai is the name of the UI

[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaas-ui/README.md:94]: The UI is called Kai.

File: gaas-ui/src/rest/clients/rest-client.ts [hunk 5]
```diff
@@ -96,18 +110,36 @@ export class RestClient<T> {
...
+                        withCredentials: true,
+                    });
+                } else {
+                    response = await axios({
+                        baseURL: restClient.baseURL,
+                        url: restClient.url,
+                        method: restClient.method,
+                        headers: restClient.headers,
+                        data: restClient.data,
+                    });
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaas-ui/src/rest/clients/rest-client.ts:140]: the withCredentials config only seems to apply to connecting to OpenShift at the moment.

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/factories/GafferFactoryTest.java [hunk 6]
```diff
@@ -93,46 +96,37 @@ public void federatedStoreRequest_shouldReturnFederatedRequestBodyWithAllOperati
...
...affer\",\"metadata\":{\"labels\":{\"configName\":\"federatedOpDeclaration\"},\"name\":\"MyGraph\"},\"spec\":{\"graph\":{\"storeProperties\":{\"gaffer.serialiser.json.modules\":\"uk.gov.gchq.gaffer.sketches.serialisation.json.SketchesJsonModules\",\"gaffer.store.properties.class\":\"uk.gov.gchq.gaffer.federatedstore.FederatedStoreProperties\",\"gaffer.store.class\":\"uk.gov.gchq.gaffer.federatedstore.FederatedStore\"},\"operationDeclarations\":[{\"handler\":{\"class\":\"uk.gov.gchq.gaffer.proxystore.operation.handler.GetProxyUrlHandler\"},\"operation\":\"uk.gov.gchq.gaffer.proxystore.operation.GetProxyUrl\"},{\"handler\":{\"class\":\"uk.gov.gchq.gaffer.proxystore.operation.handler.TestOperationHandler\"},\"operation\":\"uk.gov.gchq.gaffer.proxystore.operation.TestOperation\"}],\"config\":{\"configName\":\"federatedOpDeclaration\",\"description\":\"Another description\",\"graphId\":\"MyGraph\",\"hooks\":[{\"class\":\"uk.gov.gchq.gaffer.graph.hook.OperationAuthoriser\",\"auths\":{\"uk.gov.gchq.gaffer.federatedstore.operation.AddGraph\":[\"GAAS_SYSTEM_USER\"]}}]}},\"ingress\":{\"host\":\"mygraph-kai-dev.apps.my.kubernetes.cluster\",\"pathPrefix\":{\"ui\":\"/ui\",\"api\":\"/rest\"}}}}";
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/factories/GafferFactoryTest.java]: done

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java (hunk 3 above)
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java:41]: The fabric8 kubernetesClient is widely used and is still supported/actively updated. Unfortunately the official API is rather limited in its functionality so would not support all the things we want to do.

File: gaffer-as-a-service/pom.xml (hunk 2 above)
Comment (rocky341) [reply gaffer-as-a-service/pom.xml]: Think that this must have been a mistake while merging. Have resolved

File: gaas-ui/test/components/navigation/navigation-appbar.test.tsx [hunk 7]
```diff
@@ -55,39 +55,73 @@ describe("Navigation Appbar Component", () => {
...
-    });
-    it("should display the non-email username of the User who signed in", () => {
-        mockAuthClient();
-        inputUsername("Batman");
-        inputPassword("zxcvb");
+            expect(component.find("div#signedin-user-details").text()).toBe("HARRYHarry@gmail.com");
+        });
+        it("should display the non-email username of the User who signed in", () => {
+            mockAuthClient();
+            inputUsername("Batman");
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaas-ui/test/components/navigation/navigation-appbar.test.tsx]: fixed

File: gaffer-as-a-service/gaas-rest/deploy/helm/templates/deployment.yaml [hunk 8]
```diff
@@ -34,4 +34,11 @@ spec:
...
-                  name: {{ include "helm.fullname" . }}
\ No newline at end of file
+                  name: {{ include "helm.fullname" . }}
+            resources:
+              requests:
+                memory: "1Gi"
+                cpu: "1000m"
+              limits:
+                memory: "1Gi"
+                cpu: "1000m"
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/deploy/helm/templates/deployment.yaml]: fixed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/HelmCommand.java [hunk 9]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright 2020 Crown Copyright
+ * Copyright 2021 Crown Copyright
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/HelmCommand.java]: fixed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtil.java [hunk 10]
```diff
@@ -35,7 +36,7 @@
 
     private static final long serialVersionUID = -2550185165626007488L;
 
-    public static final long JWT_TOKEN_VALIDITY = 5 * 60 * 60;
+    public static final long JWT_TOKEN_VALIDITY = 18000;
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtil.java]: fixed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/GafferFactory.java [hunk 11]
```diff
@@ -163,9 +173,9 @@ private static GafferSpec overrideGafferSpecConfig(final GafferSpec config, fina
             final List<String> result = new ArrayList();
             if (exsistingAuths instanceof List) {
                 List<String> authsResult = (ArrayList) exsistingAuths;
-                for (int i = 0; i < authsResult.size(); i++) {
-                    if (authsResult.get(i) != "" && authsResult.get(i) != "null" && authsResult.get(i) != null) {
-                        result.add(authsResult.get(i));
+                for (final String auth : authsResult) {
+                    if (!Objects.equals(auth, "") && !Objects.equals(auth, "null") && auth != null) {
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/factories/GafferFactory.java]: fixed

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java [hunk 12]
```diff
@@ -0,0 +1,309 @@
+/*
+ * Copyright 2020 Crown Copyright
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/handlers/DeploymentHandler.java]: fixed

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtilTest.java [hunk 13]
```diff
@@ -0,0 +1 @@
...
        given(intruder.getUsername()).willReturn("n3nad");

        //when
        String token = jwtTokenUtil.generateToken(authenticatedUser);
        boolean tokenValidated = jwtTokenUtil.validateToken(token, intruder);

        //then
        assertThat(tokenValidated).isFalse();
    }
}
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtilTest.java:1]: checked in IDE and looks fine

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/client/graph/ValidateGraphHostOperationTest.java [hunk 14]
```diff
@@ -127,7 +129,7 @@ public void shouldThrowGraphNotUpException_WhenGraphReturnsDownStatus() throws I
     }
 
     @Test
-    public void whenResponseIsSuccessfulAndGraphIsUp_ShouldNotThrowException() throws InterruptedException {
```
[the remark this answers is not in the document]
Comment (K-itKat20) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/client/graph/ValidateGraphHostOperationTest.java:130]: tests don't need to be public to run, think all the tests now have public removed.

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/auth/WebSecurityConfig.java [hunk 15]
```diff
@@ -108,6 +116,9 @@ protected void configure(final HttpSecurity http) throws Exception {
             // Add a filter to validate the tokens with every request
             http.addFilterBefore(jwtRequestFilter, UsernamePasswordAuthenticationFilter.class);
         }
+        if (openshiftEnabled) {
+            http.csrf().disable().authorizeRequests().antMatchers("/**").permitAll(); // nosemgrep: java.spring.security.audit.spring-csrf-disabled.spring-csrf-disabled
```
[the remark this answers is not in the document]
Comment (rocky341) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/auth/WebSecurityConfig.java:120]: Openshift related changes

**Comment by rocky341** [comment]
@t92549 
Is this ready to merge?

File: gaas-ui/src/components/navigation-bar/AppRoutes.tsx [hunk 16]
```diff
@@ -14,11 +13,6 @@ const AppRoutes = [
         sidebarName: "View Graphs",
         component: ViewGraph,
     },
-    {
-        path: "/namespaces",
-        sidebarName: "Cluster Namespaces",
-        component: ClusterNamespaces,
```
[the remark this answers is not in the document]
Comment (t92549) [reply gaas-ui/src/components/navigation-bar/AppRoutes.tsx:20]: Removed as no longer needed

File: gaas-ui/test/components/navigation/navigation-appbar.test.tsx (hunk 7 above)
[the remark this answers is not in the document]
Comment (t92549) [reply gaas-ui/test/components/navigation/navigation-appbar.test.tsx]: I think it still says Batman, again I know it's tedious but probably best to use something generic

[the remark this answers is not in the document]
Comment (t92549) [reply gaas-ui/test/components/navigation/navigation-appbar.test.tsx]: Never mind my files hadn't updated, thank you!

File: gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java (hunk 3 above)
Comment (t92549) [reply gaffer-as-a-service/gaas-rest/src/main/java/uk/gov/gchq/gaffer/gaas/client/GafferClient.java:41]: Combining the UI and CRD makes the whole thing less complex and easier to install

File: gaas-ui/src/rest/clients/rest-client.ts (hunk 5 above)
[the remark this answers is not in the document]
Comment (rocky341) [reply gaas-ui/src/rest/clients/rest-client.ts:140]: OpenShift related change

File: gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtilTest.java (hunk 13 above)
[the remark this answers is not in the document]
Comment (t92549) [reply gaffer-as-a-service/gaas-rest/src/test/java/uk/gov/gchq/gaffer/gaas/auth/JwtTokenUtilTest.java:1]: It does look okay but I think it has strange line endings which can be fixed to make it render better on GitHub

**Comment by rocky341** [comment]
@t92549 
Yeah ready to merge. Not sure what's going on with that file I have looked at it and it looks fine in the IDE. The copyright section is identical to merge.

**Review by t92549 — APPROVED** [review]

**Pull request merged**